message_view: Add `resolve topic` button in message recipient bar.

akshatdalton opened the pull request:
Now, one can (un)resolve the topic from 2 places:

1. Stream popovers.
2. Message recipient bars.

Fixes: #18988.

timabbott commented on static/js/click_handlers.js:
```diff
@@ -340,6 +340,19 @@ export function initialize() {
         row.find(".markdown_preview").show();
     });
 
+    // TOPIC RESOLVING
```
Let's call this "RESOLVED TOPICS"; that'll be more greppable.

timabbott commented on static/js/message_edit.js:
```diff
@@ -593,6 +595,30 @@ export function start(row, edit_box_open_callback) {
+        url: "/json/messages/" + message.id,
+        data: request,
+    });
+}
```
This is great, but we probably want to do the refactoring to make this share code with the `stream_popovers.js` logic.
I think structurally, if this function accepted a `message_id` and a topic, we could have the `stream_popover.js` code call it?

timabbott replied:
(Maybe it makes sense to do this as a series of preparatory refactoring commits).

timabbott commented on static/js/message_list_view.js:
```diff
@@ -34,6 +34,8 @@ import * as submessage from "./submessage";
 import * as timerender from "./timerender";
 import * as util from "./util";
 
+const RESOLVED_TOPIC_PREFIX = "✔ ";
```
Let's do a preparatory refactoring commit moving this to be exported as `message_edit.RESOLVED_TOPIC_PREFIX`, so that we don't end up with multiple copies of this definition.

timabbott commented on static/templates/recipient_row.hbs:
```diff
@@ -50,6 +50,12 @@
+                {{else}}
+                    <i class="fa fa-check on_hover_topic_resolve recipient_bar_icon" title="{{t 'Resolve topic' }}" role="button" tabindex="0" aria-label="{{t 'Resolve topic' }}"></i>
+                {{/if}}
+
```
After playing with this, I think we want to migrate all of the `recipient_bar_icon` elements to use TippyJS tooltips for consistency with the rest of the message view UI.  Feel free to do that in a follow-up PR or extra commits in this PR series.

timabbott replied:
Also, the labels should match what we use in the popover: "Mark topic resolved" (or whatever it was).  It's very important to use consistent terminology for a feature in user-facing UI.

akshatdalton commented:
The current push addresses the following changes:
1. Changed `TOPIC RESOLVING` TO `RESOLVED TOPICS`. https://github.com/zulip/zulip/pull/19011#discussion_r659095257
2. Added `handle_resolve_topic` helper function in `message_edit.js` so that it can be used by `stream_popovers.js`. https://github.com/zulip/zulip/pull/19011#discussion_r659095406
3. Extracted `RESOLVED_TOPIC_PREFIX` in `message_edit.js`. https://github.com/zulip/zulip/pull/19011#discussion_r659095528
4. Migrated all of the `recipient_bar` icon elements to use TippyJS tooltips. https://github.com/zulip/zulip/pull/19011#discussion_r659095665
5. Corrected the label in the recipient bar to match that of stream popovers. https://github.com/zulip/zulip/pull/19011#discussion_r659095757
---
After adding TippyJS to the recipient bars, results were not quite convincing:
![Screenshot from 2021-06-26 23-26-24](https://user-images.githubusercontent.com/67277428/123523089-3169cb00-d6df-11eb-8e46-88ca2e93d873.png)
![Screenshot from 2021-06-26 23-28-52](https://user-images.githubusercontent.com/67277428/123523093-34fd5200-d6df-11eb-8e9a-419906ed6fe3.png)
In the first image, some part of the tooltip is hidden. And in the 2nd image, the tooltip is completely hidden. (Tooltip position, in this case, is set to `bottom`)

If we want to use the tooltip with the position set to `right`, then we can improve it by:
1. changing the tooltip font/box dimension just for this case.
2. or let's leave it for this case. (i.e. use HTML title) :man_shrugging: 
I have not looked properly for the solution for the tooltip with the position set to `bottom`.

Also, can you also please tell me how do we use the attribute `hidden-for-spectators`?
When I logged in as Aaron, I had the following situations when I (un)muted a topic:
![Screenshot from 2021-06-26 23-38-29](https://user-images.githubusercontent.com/67277428/123523327-8b1ec500-d6e0-11eb-9d17-6f9f09ec9c69.png)
![Screenshot from 2021-06-26 23-38-37](https://user-images.githubusercontent.com/67277428/123523329-8e19b580-d6e0-11eb-8951-9b048884d73b.png)

timabbott commented:
We have more or less standardized on `placement: top` for tooltips on most elements; we should use it here.

akshatdalton commented:
> We have more or less standardized on `placement: top` for tooltips on most elements; we should use it here.

It is not working here: I mean tooltip hides behind the message box.

timabbott commented:
Oh, I see -- we had that problem with the reactions tooltips and solved it by changing the outer container for the element.  @amanagr can help you figure out how to do that properly.

timabbott commented:
I merged a couple commits as the series ending with a7681d4a1701d4ffe11032c939d4c56c55b1d46a, after renaming the extracted function to `toggle_resolve_topic`, for more consistent naming.  

What remains looks good aside from the tooltip positioning issue.

timabbott commented:
@akshatdalton ping on this; do you need help figuring out the tooltip stuff?  If so, a thread in #frontend would be a good idea.

Probably worth rebasing before asking for help so it's easy for other folks to run this.

akshatdalton commented:
@timabbott, this is ready for review. Please try the changes.

timabbott commented:
This is great, merged, thanks @akshatdalton!
